c++/nda/h5.hpp

```diff
@@ -155,28 +210,47 @@ namespace nda {
         return;
       }
 
-      int rank_in_file = lt.rank() - (is_complex ? 1 : 0);
-      if (rank_in_file != A::rank)
-        NDA_RUNTIME_ERROR << " h5 read of nda::array : incorrect rank. In file: " << rank_in_file << "  In memory " << A::rank;
-      std::array<long, A::rank> L;
-      for (int u = 0; u < A::rank; ++u) L[u] = lt.lengths[u]; // NB : correct for complex
+      int rank_in_file = lt.rank() - is_complex;
+      std::array<long, A::rank> shape;
+      auto slice_slab = h5::array_interface::hyperslab{};
+
+      if constexpr (slicing) {
+        if (rank_in_file != size_of_slice)
+          NDA_RUNTIME_ERROR << " h5 read of nda::array : incorrect slice rank. In file: " << rank_in_file << "  Rank of slice: " << size_of_slice;
+        auto const [sl, sh] = hyperslab_and_shape_from_slice(slice, lt.lengths, is_complex);
+        static_assert(std::tuple_size_v<decltype(sh)> == A::rank, "Array rank does not match the number of non-trivial slice dimensions");
+        slice_slab = sl;
+        shape      = sh;
+      } else {
+        if (rank_in_file != A::rank)
+          NDA_RUNTIME_ERROR << " h5 read of nda::array : incorrect rank. In file: " << rank_in_file << "  In memory: " << A::rank;
+        for (int u = 0; u < A::rank; ++u) shape[u] = lt.lengths[u]; // NB : correct for complex
+      }
 
       if constexpr (is_regular_v<A>) {
-        a.resize(L);
+        a.resize(shape);
       } else {
-        if (a.shape() != L)
+        if (a.shape() != shape)
           NDA_RUNTIME_ERROR << "Error trying to read from an hdf5 file to a view. Dimension mismatch"
-                            << "\n in file  : " << L << "\n in view  : " << a.shape();
+                            << "\n in file  : " << shape << "\n in view  : " << a.shape();
       }
 
-      h5::array_interface::h5_array_view v{h5::hdf5_type<get_value_t<A>>(), (void *)(a.data()), A::rank, is_complex};
-      for (int u = 0; u < A::rank; ++u) {
-        v.slab.count[u] = L[u];
-        v.L_tot[u]      = L[u];
+      h5::array_interface::h5_array_view v{h5::hdf5_type<get_value_t<A>>(), (void *)(a.data()), rank_in_file, is_complex};
+      if constexpr (slicing) {
+        v.slab.count = slice_slab.count;
+        v.L_tot      = slice_slab.count;
+      } else {
+        for (int u = 0; u < A::rank; ++u) {
+          v.slab.count[u] = shape[u];
+          v.L_tot[u]      = shape[u];
+        }
       }
-      h5::array_interface::read(g, name, v, lt);
+      h5::array_interface::read(g, name, v, lt, slice_slab);
 
     } else { // generic unknown type to hdf5
+
+      static_assert(!slicing, "Slicing not supported in generic h5_read");
+
       auto g2 = g.open_group(name);
```

Comment:
To be discussed.
Should we split the function into an impl and a function handles the slab template ?
Or use slab{} type ? 
In the past, templating h5_read a lot lead to long compilation time at the end ...

Reply:
The new functionality should not effect the number of instantiations for those cases where we don't slice.
Slicing will not be the common use-case so I don't foresee a problem here.